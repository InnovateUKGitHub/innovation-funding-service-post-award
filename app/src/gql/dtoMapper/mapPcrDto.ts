import { TypeOfAid } from "@framework/constants/project";
import { ProjectChangeRequest } from "@framework/constants/recordTypes";
import { PCRDto, FullPCRItemDto } from "@framework/dtos/pcrDtos";
import {
  getPCROrganisationType,
  mapToPCRItemStatus,
  mapFromSalesforcePCRProjectRole,
  mapTypeOfAidToEnum,
  mapToPCRStatus,
  mapFromSalesforcePCRPartnerType,
  mapToPcrItemType,
} from "@framework/mappers/pcr";
import { pcrRecordTypeMetaValues } from "@framework/mappers/pcrRecordTypeMetaData";
import { Clock } from "@framework/util/clock";

// on Project_Change_Requests__r

const clock = new Clock();

export type PcrNode = Readonly<
  Partial<{
    Id: string;
    Acc_AdditionalNumberofMonths__c: GQL.Value<number>;
    Acc_Comments__c: GQL.Value<string>;
    Acc_ExistingPartnerName__c: GQL.Value<string>;
    Acc_OtherFunding__c: GQL.Value<boolean>;
    Acc_CommercialWork__c: GQL.Value<boolean>;
    Acc_ExistingProjectDuration__c: GQL.Value<number>;
    Acc_NewOrganisationName__c: GQL.Value<string>;
    Acc_Reasoning__c: GQL.Value<string>;
    Acc_RequestHeader__c: GQL.Value<string>;
    Acc_RequestNumber__c: GQL.Value<number>;
    Acc_MarkedasComplete__c: GQL.Value<string>;
    Acc_OrganisationName__c: GQL.Value<string>;
    Acc_ParticipantType__c: GQL.Value<string>;

    CreatedDate: GQL.Value<string>;
    Acc_Status__c: GQL.Value<string>;
    Acc_Project__c: GQL.Value<string>;
    Acc_ProjectRole__c: GQL.Value<string>;
    LastModifiedDate: GQL.Value<string>;
    RecordType: {
      DeveloperName?: GQL.Value<string>;
      Name?: GQL.ValueAndLabel<string>;
      Id?: string | null;
    } | null;
  }>
> | null;

type PcrDtoMapping = Pick<
  PCRDto,
  | "comments"
  | "id"
  | "lastUpdated"
  | "projectId"
  | "reasoningComments"
  | "reasoningStatus"
  | "requestNumber"
  | "started"
  | "status"
  | "statusName"
>;

export type PcrItemDtoMapping = Pick<
  FullPCRItemDto,
  | "accountName"
  | "hasOtherFunding"
  | "id"
  | "isCommercialWork"
  | "guidance"
  | "lastUpdated"
  | "offsetMonths"
  | "organisationName"
  | "organisationType"
  | "partnerNameSnapshot"
  | "partnerType"
  | "projectDurationSnapshot"
  | "projectId"
  | "projectRole"
  | "requestNumber"
  | "shortName"
  | "started"
  | "status"
  | "statusName"
  | "type"
  | "typeName"
  | "typeOfAid"
>;

/**
 * Mapper for PCR Child items
 */
const itemMapper: GQL.DtoMapper<PcrItemDtoMapping, PcrNode, { typeOfAid?: string | TypeOfAid }> = {
  accountName(node) {
    return node?.Acc_NewOrganisationName__c?.value ?? null;
  },
  hasOtherFunding(node) {
    return node?.Acc_OtherFunding__c?.value ?? null;
  },
  id(node) {
    return node?.Id as PcrItemId;
  },
  isCommercialWork(node) {
    return node?.Acc_CommercialWork__c?.value ?? null;
  },
  guidance(node, additionalData) {
    const metaValues = pcrRecordTypeMetaValues.find(x => x.type === itemMapper.type(node, additionalData));
    if (!!metaValues && "guidance" in metaValues) {
      return metaValues.guidance;
    }
    return "";
  },
  lastUpdated(node) {
    return node?.LastModifiedDate?.value
      ? clock.parseRequiredSalesforceDateTime(node?.LastModifiedDate?.value)
      : new Date();
  },
  offsetMonths(node) {
    return node?.Acc_AdditionalNumberofMonths__c?.value ?? 0;
  },
  organisationName(node) {
    return node?.Acc_OrganisationName__c?.value ?? null;
  },
  organisationType(node) {
    return getPCROrganisationType(mapFromSalesforcePCRPartnerType(node?.Acc_ParticipantType__c?.value ?? ""));
  },
  partnerNameSnapshot(node) {
    return node?.Acc_ExistingPartnerName__c?.value ?? null;
  },
  partnerType(node) {
    return mapFromSalesforcePCRPartnerType(node?.Acc_ParticipantType__c?.value ?? "");
  },
  projectDurationSnapshot(node) {
    return node?.Acc_ExistingProjectDuration__c?.value ?? 0;
  },
  projectId(node) {
    return (node?.Acc_Project__c?.value ?? "unknown-project-id") as ProjectId;
  },
  projectRole(node) {
    return mapFromSalesforcePCRProjectRole(node?.Acc_ProjectRole__c?.value ?? "");
  },
  requestNumber(node) {
    return node?.Acc_RequestNumber__c?.value ?? 0;
  },
  shortName(node) {
    return node?.RecordType?.Name?.value ?? "Unknown";
  },
  started(node) {
    return node?.CreatedDate?.value ? clock.parseRequiredSalesforceDateTime(node?.CreatedDate?.value) : new Date();
  },
  status(node) {
    return mapToPCRItemStatus(node?.Acc_MarkedasComplete__c?.value ?? "");
  },
  statusName(node) {
    return node?.Acc_MarkedasComplete__c?.value || "Unknown";
  },
  type(node) {
    return mapToPcrItemType(node?.RecordType?.DeveloperName?.value ?? node?.RecordType?.Name?.value ?? "Unknown");
  },
  typeName(node) {
    return node?.RecordType?.Name?.value ?? "Unknown";
  },
  typeOfAid(node, additionalData) {
    /* possible to be passed in as TypeofAid enum from project mapper or as string from salesforce */
    return additionalData?.typeOfAid === undefined
      ? TypeOfAid.Unknown
      : typeof additionalData.typeOfAid === "string"
      ? mapTypeOfAidToEnum(additionalData?.typeOfAid)
      : additionalData.typeOfAid;
  },
};

/**
 * Mapper for the PCR header node
 */
const headMapper: GQL.DtoMapper<PcrDtoMapping, PcrNode> = {
  comments(node) {
    return node?.Acc_Comments__c?.value ?? "";
  },
  id(node) {
    return (node?.Id ?? "") as PcrId;
  },
  lastUpdated(node) {
    return node?.LastModifiedDate?.value
      ? clock.parseRequiredSalesforceDateTime(node?.LastModifiedDate?.value)
      : new Date();
  },
  projectId(node) {
    return (node?.Acc_Project__c?.value ?? "unknown-project-id") as ProjectId;
  },
  reasoningComments(node) {
    return node?.Acc_Reasoning__c?.value ?? "";
  },
  reasoningStatus(node) {
    return mapToPCRItemStatus(node?.Acc_MarkedasComplete__c?.value ?? "");
  },
  requestNumber(node) {
    return node?.Acc_RequestNumber__c?.value ?? 0;
  },
  started(node) {
    return node?.CreatedDate?.value ? clock.parseRequiredSalesforceDateTime(node?.CreatedDate?.value) : new Date();
  },
  status(node) {
    return mapToPCRStatus(node?.Acc_Status__c?.value || "unknown");
  },
  statusName(node) {
    return node?.Acc_Status__c?.value || "Unknown";
  },
};

type CollatedPcrNode<T extends PcrNode = PcrNode> = {
  head: T;
  children: T[];
};

type PcrDtoWithItems<PickList extends keyof PcrDtoMapping, ItemPickList extends keyof PcrItemDtoMapping> = Merge<
  Pick<PcrDtoMapping, PickList>,
  { items: Pick<PcrItemDtoMapping, ItemPickList>[] }
>;

type PcrAdditionalData<TPickList extends string> = AdditionalDataType<
  TPickList,
  [
    ["typeOfAid", "typeOfAid", string | TypeOfAid], // get from Acc_CompetitionId__r { Acc_TypeofAid__c {value}} or from project.typeOfAid
  ]
>;

/**
 * maps for a single pcr item
 */
export function mapPcrItemDto<T extends PcrNode, PickList extends keyof PcrItemDtoMapping>(
  node: T,
  pickList: PickList[],
  additionalData: PcrAdditionalData<PickList>,
): Pick<PcrItemDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = itemMapper[field](node, additionalData);
    return dto;
  }, {} as Pick<PcrItemDtoMapping, PickList>);
}

/**
 * maps the attached child PCR items to the dto pattern
 */
export function mapPcrItemsDtos<ItemPickList extends keyof PcrItemDtoMapping>(
  items: CollatedPcrNode["children"],
  pickList: ItemPickList[],
  additionalData: PcrAdditionalData<ItemPickList>,
) {
  return items.map(item =>
    pickList.reduce((dto, field) => {
      dto[field] = itemMapper[field](item, additionalData);
      return dto;
    }, {} as Pick<PcrItemDtoMapping, ItemPickList>),
  );
}

/**
 * maps the header of the pcr to the dto and passes the children to the item mapper
 */
export function mapToPcrDto<
  T extends CollatedPcrNode,
  PickList extends keyof PcrDtoMapping,
  ItemPickList extends keyof PcrItemDtoMapping,
>(
  pcr: T,
  pickList: PickList[],
  itemsPickList: ItemPickList[],
  additionalData: PcrAdditionalData<ItemPickList>,
): PcrDtoWithItems<PickList, ItemPickList> {
  const dtoWithoutItems = pickList.reduce((dto, field) => {
    {
      dto[field] = headMapper[field](pcr.head);
    }
    return dto;
  }, {} as Pick<PcrDtoMapping, PickList>);

  const dtoWithItems = {
    ...dtoWithoutItems,
    items: mapPcrItemsDtos(pcr.children, itemsPickList, additionalData),
  } as PcrDtoWithItems<PickList, ItemPickList>;
  return dtoWithItems;
}

/**
 * Separates the pcrs into headers linked with attached child pcrs and sends to mapper
 */
export function mapToPcrDtoArray<
  T extends PcrNode,
  PickList extends keyof PcrDtoMapping,
  ItemPickList extends keyof PcrItemDtoMapping,
>(
  pcrs: ReadonlyArray<Readonly<{ node: T | null }> | null>,
  pickList: PickList[],
  itemsPickList: ItemPickList[],
  additionalData: PcrAdditionalData<ItemPickList>,
): PcrDtoWithItems<PickList, ItemPickList>[] {
  if (!pcrs) return [];

  const collatedPcrs = [];

  for (const pcr of pcrs) {
    if (typeof pcr === null) continue;
    if (pcr?.node?.RecordType?.DeveloperName?.value === ProjectChangeRequest.requestHeader) {
      const childPcrs = [];

      for (const childPcr of pcrs) {
        if (!!pcr?.node?.Id && pcr?.node?.Id === childPcr?.node?.Acc_RequestHeader__c?.value) {
          childPcrs.push(childPcr);
        }
      }

      collatedPcrs.push({
        head: pcr?.node,
        children: childPcrs.map(x => x?.node) as T[],
      });
    }
  }

  return collatedPcrs.map(pcr =>
    mapToPcrDto<CollatedPcrNode<T>, PickList, ItemPickList>(pcr, pickList, itemsPickList, additionalData),
  );
}
