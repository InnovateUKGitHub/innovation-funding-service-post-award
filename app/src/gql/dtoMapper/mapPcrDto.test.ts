import { mapToPcrDtoArray, PcrNode } from "./mapPcrDto";

const edges = [
  {
    node: {
      Id: "a0G26000007K4vtEAC",
      Acc_Status__c: {
        value: null,
      },
      Acc_RequestHeader__c: {
        value: "a0G26000007K4veEAC",
      },
      Acc_RequestNumber__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-04-13T09:36:55.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-13T09:38:24.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Change project scope",
          label: "Change project scope",
        },
      },
    },
  },
  {
    node: {
      Id: "a0G26000007KUl6EAG",
      Acc_Status__c: {
        value: null,
      },
      Acc_RequestHeader__c: {
        value: "a0G26000007KUl3EAG",
      },
      Acc_RequestNumber__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Change a partner's name",
          label: "Change a partner's name",
        },
      },
    },
  },
  {
    node: {
      Id: "a0G26000007KUl5EAG",
      Acc_Status__c: {
        value: null,
      },
      Acc_RequestHeader__c: {
        value: "a0G26000007KUl3EAG",
      },
      Acc_RequestNumber__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Add a partner",
          label: "Add a partner",
        },
      },
    },
  },
  {
    node: {
      Id: "a0G26000007KUl4EAG",
      Acc_Status__c: {
        value: null,
      },
      Acc_RequestHeader__c: {
        value: "a0G26000007KUl3EAG",
      },
      Acc_RequestNumber__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-19T10:17:21.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Reallocate several partners' project cost",
          label: "Reallocate several partners' project cost",
        },
      },
    },
  },
  {
    node: {
      Id: "a0G26000007KUl3EAG",
      Acc_Status__c: {
        value: "Draft",
      },
      Acc_RequestHeader__c: {
        value: null,
      },
      Acc_RequestNumber__c: {
        value: 3,
      },
      CreatedDate: {
        value: "2023-04-19T10:17:19.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-19T10:17:24.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Request Header",
          label: "Request Header",
        },
      },
    },
  },
  {
    node: {
      Id: "a0G26000007K4veEAC",
      Acc_Status__c: {
        value: "Submitted to Monitoring Officer",
      },
      Acc_RequestHeader__c: {
        value: null,
      },
      Acc_RequestNumber__c: {
        value: 2,
      },
      CreatedDate: {
        value: "2023-04-13T09:36:53.000Z",
      },
      LastModifiedDate: {
        value: "2023-04-13T09:39:06.000Z",
      },
      Acc_Project__c: {
        value: "a0E2600000o75OAEAY",
      },
      RecordType: {
        Name: {
          value: "Request Header",
          label: "Request Header",
        },
      },
    },
  },
] as ReadonlyArray<Readonly<{ node: PcrNode } | null> | null>;

describe("mapPcrDtoArray", () => {
  it("should map the node to the projectDto structure", () => {
    expect(
      mapToPcrDtoArray(
        edges,
        ["status", "statusName", "started", "lastUpdated", "id", "projectId", "requestNumber"],
        ["shortName"],
      ),
    ).toMatchSnapshot();
  });
});
