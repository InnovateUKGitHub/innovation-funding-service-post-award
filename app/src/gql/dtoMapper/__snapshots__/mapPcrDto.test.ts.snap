// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPcrDtoArray should map the node to the projectDto structure 1`] = `
Array [
  Object {
    "id": "a0G26000007KkbiEAC",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KkbnEAC",
        "isCommercialWork": false,
        "organisationName": "AAA LTD",
        "organisationType": "Industrial",
        "partnerNameSnapshot": null,
        "partnerType": 10,
        "projectRole": 20,
        "shortName": "Add a partner",
        "status": 3,
        "type": 20,
        "typeName": "Add a partner",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-05-24T09:53:31.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 3,
    "requestNumber": 7,
    "started": 2023-05-10T13:55:22.000Z,
    "status": 10,
    "statusName": "Approved",
  },
  Object {
    "id": "a0G26000007KkfzEAC",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007Kkg4EAC",
        "isCommercialWork": false,
        "organisationName": null,
        "organisationType": "",
        "partnerNameSnapshot": null,
        "partnerType": 0,
        "projectRole": 20,
        "shortName": "Change project scope",
        "status": 3,
        "type": 80,
        "typeName": "Change project scope",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-05-17T15:11:14.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 3,
    "requestNumber": 8,
    "started": 2023-05-10T15:56:40.000Z,
    "status": 9,
    "statusName": "Withdrawn",
  },
  Object {
    "id": "a0G26000007Ko6gEAC",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007Ko6lEAC",
        "isCommercialWork": false,
        "organisationName": null,
        "organisationType": "",
        "partnerNameSnapshot": null,
        "partnerType": 0,
        "projectRole": 20,
        "shortName": "Change project duration",
        "status": 3,
        "type": 90,
        "typeName": "Change project duration",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-05-18T14:57:59.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 3,
    "requestNumber": 11,
    "started": 2023-05-18T14:56:25.000Z,
    "status": 3,
    "statusName": "Queried by Monitoring Officer",
  },
  Object {
    "id": "a0G26000007KqPTEA0",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KqPYEA0",
        "isCommercialWork": true,
        "organisationName": "C",
        "organisationType": "Industrial",
        "partnerNameSnapshot": null,
        "partnerType": 10,
        "projectRole": 20,
        "shortName": "Add a partner",
        "status": 3,
        "type": 20,
        "typeName": "Add a partner",
        "typeOfAid": 0,
      },
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KqPdEAK",
        "isCommercialWork": true,
        "organisationName": "A LIMITED",
        "organisationType": "Industrial",
        "partnerNameSnapshot": null,
        "partnerType": 10,
        "projectRole": 20,
        "shortName": "Add a partner",
        "status": 3,
        "type": 20,
        "typeName": "Add a partner",
        "typeOfAid": 0,
      },
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KqPiEAK",
        "isCommercialWork": true,
        "organisationName": "D LIMITED",
        "organisationType": "Industrial",
        "partnerNameSnapshot": null,
        "partnerType": 10,
        "projectRole": 20,
        "shortName": "Add a partner",
        "status": 3,
        "type": 20,
        "typeName": "Add a partner",
        "typeOfAid": 0,
      },
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KqPsEAK",
        "isCommercialWork": false,
        "organisationName": null,
        "organisationType": "",
        "partnerNameSnapshot": "A B Cad Services",
        "partnerType": 0,
        "projectRole": 20,
        "shortName": "Remove a partner",
        "status": 3,
        "type": 30,
        "typeName": "Remove a partner",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-05-24T09:46:08.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 3,
    "requestNumber": 14,
    "started": 2023-05-24T09:29:47.000Z,
    "status": 12,
    "statusName": "Submitted to Innovate UK",
  },
  Object {
    "id": "a0G26000007KtLnEAK",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KtLsEAK",
        "isCommercialWork": false,
        "organisationName": null,
        "organisationType": "",
        "partnerNameSnapshot": null,
        "partnerType": 0,
        "projectRole": 20,
        "shortName": "Reallocate several partners' project cost",
        "status": 1,
        "type": 60,
        "typeName": "Reallocate several partners' project cost",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-06-02T09:43:53.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 1,
    "requestNumber": 21,
    "started": 2023-06-02T09:43:50.000Z,
    "status": 1,
    "statusName": "Draft",
  },
  Object {
    "id": "a0G26000007KtsoEAC",
    "items": Array [
      Object {
        "accountName": null,
        "hasOtherFunding": false,
        "id": "a0G26000007KtstEAC",
        "isCommercialWork": false,
        "organisationName": null,
        "organisationType": "",
        "partnerNameSnapshot": null,
        "partnerType": 0,
        "projectRole": 0,
        "shortName": "Add a partner",
        "status": 1,
        "type": 20,
        "typeName": "Add a partner",
        "typeOfAid": 0,
      },
    ],
    "lastUpdated": 2023-06-05T10:53:38.000Z,
    "projectId": "a0E2600000okzO9EAI",
    "reasoningStatus": 1,
    "requestNumber": 22,
    "started": 2023-06-05T10:53:36.000Z,
    "status": 1,
    "statusName": "Draft",
  },
]
`;
